Bluetooth: Controller: Fix max PDU len radio config for BIS Ctrl PDUs

cvinayak opened the pull request:
Fix missing radio configuration that limits the maximum PDU
len on reception to consider that Control PDU can be larger
than the max PDU size in the BIS parameters.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix max PDU len radio config for BIS Ctrl PDUs

Fix missing radio configuration that limits the maximum PDU
len on reception to consider that Control PDU can be larger
than the max PDU size in the BIS parameters.
Use dedicated PDU buffer for transmission and reception of
BIG Control PDUs so that it is independent of maximum ISO
PDU length which can be less than BIG Control PDU length.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: Remove duplicate board conf for iso_broadcast

Remove duplicated board specific conf file for iso_broadcast
and iso_receive samples. Instead, have an overlay conf file.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on samples/bluetooth/iso_broadcast/overlay-bt_ll_sw_split.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_BT_CTLR_ADV_ISO=y
+CONFIG_BT_CTLR_ADV_ISO_PDU_LEN_MAX=4
+CONFIG_BT_CTLR_ADV_ISO_STREAM_MAX=2
+
+CONFIG_BT_CTLR_ISOAL_SOURCES=2
+
+CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE=8
+CONFIG_BT_CTLR_ISO_TX_BUFFERS=4
```
How do users tell when to use this overlay, and when not to use it?

Thalley commented on samples/bluetooth/iso_broadcast/prj.conf:
```diff
@@ -4,9 +4,7 @@ CONFIG_BT_DEBUG_LOG=y
 
 CONFIG_BT_ISO_MAX_CHAN=2
-CONFIG_BT_ISO_TX_BUF_COUNT=2
+CONFIG_BT_ISO_TX_BUF_COUNT=4
```
Should `BT_CTLR_ISO_TX_BUFFERS` inherit the value of `BT_ISO_TX_BUF_COUNT` by default?

Thalley commented on samples/bluetooth/iso_broadcast/sample.yaml:
```diff
@@ -7,3 +7,10 @@ tests:
     integration_platforms:
       - qemu_cortex_m3
     tags: bluetooth
+  sample.bluetooth.iso_broadcast.bt_ll_sw_split:
+    harness: bluetooth
+    platform_allow: qemu_cortex_m3 qemu_x86 nrf52_bsim nrf52dk_nrf52832
+    integration_platforms:
+      - nrf52dk_nrf52832
+    extra_args: OVERLAY_CONFIG=overlay-bt_ll_sw_split.conf
```
The number of boards that had a specific config before does not match the platforms/boards that use the new overlay. Is this intentional?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -1006,20 +1024,27 @@ static void isr_rx(void *param)
 
 		payload_count = lll->payload_count - lll->bn;
 		if (bis) {
+			static uint8_t big_ctrl_pdu[offsetof(struct node_rx_pdu, pdu) +
+						    sizeof(struct pdu_big_ctrl)];
```
It's not something we do alot (but perhaps should): Using `offsetof` is a nice way to do something like this, but it is very fragile. Any changes to struct (e.g. adding or removing a field) may break the assumption here. 

One trivial approach is to of course to add a comment in the struct that this should not be modified without considering the sideeffects. 

We could probably also use a `BUILD_ASSERT` with calls to `offsetof` to help ensure correctness.

cvinayak replied:
offsetof was not needed, it was my confusion on buffer size to be allocated. Only the actual max control PDU length buffer is that all needed, now this dedicated buffer is used for both broadcaster and sync receiver.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -1006,20 +1024,27 @@ static void isr_rx(void *param)
 
 		payload_count = lll->payload_count - lll->bn;
 		if (bis) {
+			static uint8_t big_ctrl_pdu[offsetof(struct node_rx_pdu, pdu) +
```
```suggestion
			static uint8_t big_ctrl_pdu[NODE_RX_HEADER_SIZE +
```

cvinayak committed:
Bluetooth: Controller: Inherit BT_ISO_TX_BUF_COUNT value

Inherit BT_ISO_TX_BUF_COUNT value when BT_ISO is enabled
to set BT_CTLR_ISO_TX_BUFFERS count.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on samples/bluetooth/iso_broadcast/README.rst:
```diff
@@ -21,7 +21,8 @@ Building and Running
 This sample can be found under :zephyr_file:`samples/bluetooth/iso_broadcast` in
-the Zephyr tree.
+the Zephyr tree. Use `-DOVERLAY_CONFIG=overlay-bt_ll_sw_split.conf` to enable
+required ISO feature support in Zephyr Bluetooth Controller on supported boards.
```
This may be nitpicking: But how does a user determiner which board is supported? The `platform_allow` can no longer be used, as that only contains a single board (`nrf52dk_nrf52832`), but I guess this also works for the nrf52840, right?

cvinayak replied:
Good question again, myself do not know how to figure out which samples would work on which boards! I will ask the experts... and get back

cvinayak replied:
@carlescufi said 
> this is all in flux in the upstream project, there are no clear definitions

I will leave you question on, which board is supported, out of the scope of this PR.

Thalley commented on samples/bluetooth/iso_broadcast/overlay-bt_ll_sw_split.conf:
```diff
@@ -0,0 +1,10 @@
+
+CONFIG_BT_CTLR_ISOAL_SOURCES=2
+
+# FIXME: Host needs BT_ISO_TX_MTU + 4 bytes otherwise it tries to fragment ISO
+# data. When Host is fixed, BT_CTLR_ISO_TX_BUFFER_SIZE can inherit BT_ISO_TX_MTU
+# value.
```
Or 8 if using timestamp, right?

cvinayak replied:
I do not know when Host uses timestamp, but currently for this sample +4 seems to be sufficient.

Thalley replied:
This sample uses `BT_ISO_TIMESTAMP_NONE` when sending, so that will work. 

But if someone modifies the call to `bt_iso_chan_send`, then 4 octets may not be enough. 

Not something that needs to change in your PR though, but something to keep in mind

Thalley commented on samples/bluetooth/iso_broadcast/overlay-bt_ll_sw_split.conf:
```diff
@@ -0,0 +1,13 @@
+
+# FIXME: Host needs CONFIG_BT_ISO_TX_MTU + 4 bytes for sequence number, and
+# optionally additional + 4 bytes for timestamp when not using
+$ BT_ISO_TIMESTAMP_NONE in bt_iso_chan_send(), otherwise Host tries to fragment
```
```suggestion
# BT_ISO_TIMESTAMP_NONE in bt_iso_chan_send(), otherwise Host tries to fragment
```

cvinayak replied:
Fixed
